test/e2e/suite/nodeadm_test.go

```diff
@@ -348,12 +283,182 @@ var _ = Describe("Hybrid Nodes", func() {
+								nodeadmURLs:        test.nodeadmURLs,
+								os:                 os,
+								provider:           provider,
+								s3Client:           test.s3Client,
+								ssmClient:          test.ssmClient,
+							}
+							instance, err := createNodeTest.Run(ctx)
+							Expect(err).NotTo(HaveOccurred(), "EC2 Instance should have been created successfully")
+
+							joinNodeTest := joinNodeTest{
+								k8s:           test.k8sClient,
+								nodeIPAddress: instance.IP,
+								logger:        test.logger,
+							}
+							Expect(joinNodeTest.Run(ctx)).To(Succeed(), "node should have joined the cluster sucessfully")
+
+							upgradeNodeTest := upgradeNodeTest{
+								k8s:      test.k8sClient,
+								ssm:      test.ssmClient,
+								cluster:  test.cluster,
+								ec2:      instance,
+								logger:   test.logger,
+								provider: provider,
+							}
+							Expect(upgradeNodeTest.Run(ctx)).To(Succeed(), "node should have upgraded successfully")
+							Expect(joinNodeTest.Run(ctx)).To(Succeed(), "node should have joined the cluster sucessfully after nodeadm upgrade")
+
+							test.logger.Info("Resetting hybrid node...")
+
+							uninstallNodeTest := uninstallNodeTest{
+								k8s:      test.k8sClient,
+								ssm:      test.ssmClient,
+								ec2:      instance,
+								provider: provider,
+								logger:   test.logger,
+							}
+							Expect(uninstallNodeTest.Run(ctx)).To(Succeed(), "node should have been reset sucessfully")
+						},
+						Entry(fmt.Sprintf("With OS %s and with Credential Provider %s", os.Name(), string(provider.Name())), context.Background(), os, provider, Label(os.Name(), string(provider.Name()), "upgradeflow")),
+					)
 				}
 			}
 		})
 	})
 })
 
+type createNodeTest struct {
+	awsSession         *session.Session
+	cluster            *peered.HybridCluster
+	ec2ClientV2        *ec2v2.Client
+	instanceName       string
+	instanceProfileARN string
+	k8sClient          *clientgo.Clientset
+	k8sVersion         string
+	logger             logr.Logger
+	nodeadmURLs        e2e.NodeadmURLs
+	os                 e2e.NodeadmOS
+	provider           e2e.NodeadmCredentialsProvider
+	s3Client           *s3v1.S3
+	skipCleanup        bool
+	ssmClient          *ssmv1.SSM
+}
+
+func (c createNodeTest) Run(ctx context.Context) (ec2.Instance, error) {
+	Expect(c.os).NotTo(BeNil())
+	Expect(c.provider).NotTo(BeNil())
+
+	nodeSpec := e2e.NodeSpec{
+		OS: c.os,
+		Cluster: &e2e.Cluster{
+			Name:   c.cluster.Name,
+			Region: c.cluster.Region,
+		},
+		Provider: c.provider,
+	}
+
+	files, err := c.provider.FilesForNode(nodeSpec)
+	Expect(err).NotTo(HaveOccurred())
+
+	nodeadmConfig, err := c.provider.NodeadmConfig(ctx, nodeSpec)
+	Expect(err).NotTo(HaveOccurred(), "expected to build nodeconfig")
+
+	nodeadmConfigYaml, err := yaml.Marshal(&nodeadmConfig)
+	Expect(err).NotTo(HaveOccurred(), "expected to successfully marshal nodeadm config to YAML")
+
+	var rootPasswordHash string
+	if suite.TestConfig.SetRootPassword {
```

Comment:
can we pass this in the createNodeTest?
Ideally only the `BeforeEach` should be reading the `suite` variable (I think I missed this one before)
Same with the other `suite` references in this method